sklearn/neighbors/tests/test_lmnn.py

```diff
@@ -0,0 +1,486 @@
+
+    lmnn = LargeMarginNearestNeighbor(warm_start=True, max_iter=5)
+    lmnn.fit(X, y)
+
+    X_less_features, y = make_cla(n_samples=30, n_features=4, n_classes=4,
+                                  n_redundant=0, n_informative=4,
+                                  random_state=0)
+    assert_raises(ValueError, lmnn.fit, X_less_features, y)
+
+
+def test_verbose():
+    lmnn = LargeMarginNearestNeighbor(n_neighbors=3, verbose=1)
+    lmnn.fit(iris_data, iris_target)
+
+
+def test_random_state():
+    """Assert that when having more than max_impostors (forcing sampling),
+    the same impostors will be sampled given the same random_state and
+    different impostors will be sampled given a different random_state
+    leading to a different transformation"""
+
+    X = iris_data
+    y = iris_target
+    params = {'n_neighbors': 3, 'max_impostors': 5, 'random_state': 1,
+              'max_iter': 10, 'init': 'identity'}
+
+    lmnn = LargeMarginNearestNeighbor(**params)
+    lmnn.fit(X, y)
+    transformation_1 = lmnn.transformation_
+
+    lmnn = LargeMarginNearestNeighbor(**params)
+    lmnn.fit(X, y)
+    transformation_2 = lmnn.transformation_
+
+    # This assertion fails on 32bit systems if init='pca'
+    assert_allclose(transformation_1, transformation_2)
+
+    params['random_state'] = 2
+    lmnn = LargeMarginNearestNeighbor(**params)
+    lmnn.fit(X, y)
+    transformation_3 = lmnn.transformation_
+
+    assert_false(np.allclose(transformation_2, transformation_3))
+
+
+def test_singleton_class():
+    X = iris_data
+    y = iris_target
+    X_tr, X_te, y_tr, y_te = train_test_split(X, y, test_size=0.3, stratify=y)
+
+    # one singleton class
+    singleton_class = 1
+    ind_singleton, = np.where(y_tr == singleton_class)
+    y_tr[ind_singleton] = 2
+    y_tr[ind_singleton[0]] = singleton_class
+
+    lmnn = LargeMarginNearestNeighbor(n_neighbors=3, max_iter=30)
+    lmnn.fit(X_tr, y_tr)
+
+    # One non-singleton class
+    X_tr, X_te, y_tr, y_te = train_test_split(X, y, test_size=0.3, stratify=y)
+    ind_1, = np.where(y_tr == 1)
+    ind_2, = np.where(y_tr == 2)
+    y_tr[ind_1] = 0
+    y_tr[ind_1[0]] = 1
+    y_tr[ind_2] = 0
+    y_tr[ind_2[0]] = 2
+
+    lmnn = LargeMarginNearestNeighbor(n_neighbors=3, max_iter=30)
+    assert_raises(ValueError, lmnn.fit, X_tr, y_tr)
+
+
+def test_callable():
+    X = iris_data
+    y = iris_target
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)
+
+    lmnn = LargeMarginNearestNeighbor(n_neighbors=3, callback='my_cb')
+    assert_raises(ValueError, lmnn.fit, X_train, y_train)
+
+    max_iter = 10
+
+    def my_cb(transformation, n_iter):
+        rem_iter = max_iter - n_iter
+        print('{} iterations remaining...'.format(rem_iter))
+
+    lmnn = LargeMarginNearestNeighbor(n_neighbors=3, callback=my_cb,
+                                      max_iter=max_iter, verbose=1)
+    lmnn.fit(X_train, y_train)
+
+
+def test_terminate_early():
+    X = iris_data
```

Comment:
what does this test?

Reply:
do we also need to test that a ConvergenceWarning is raised if convergence criteria are not met during `fit`?

Reply:
I think this was just a sanity check that everything still works when `max_iter` is set to a small value.

Reply:
Convergence could be checked from `opt_result`. Should we keep it (?)